### PR TITLE
Switch build system to autotools

### PR DESCRIPTION
**Pull request by sedited**
This pull request will switch the current build system to autotools. Gridcoin core functions are now built into a library that is shared between the daemon and gui client. It compiles by default with Qt5 and adds some necessary features, that are needed for deterministic gitian builds. The depends system needed for gitian is not active yet, but will be completed with a future pull request. 
To build run: 
`./autogen.sh` , which runs autoreconf to generate the configure script
`./configure` , with any desired options to generate the makefile. To compile with Qt5 for your current system architecture use `./configure --with-gui=qt5`
`make` , `V=1` will enter more verbose mode, use `-j # of cores` for multithreading 

With standard configuration, the system will build both the daemon and the gui client. The daemon is placed in src/gridcoinresearchd and the gui in src/qt/gridcoinresearch. 

During compilation a library of core gridcoin functionality called libgridcoin_util is built, that is later linked with the daemon main entry point and with the rest of the qt files for the gui client. The main function of the deamon has been moved into a seperate file called bitcoind.cpp . 

Please do not merge the pull request before @Lederstrumpf has ack'ed.

Edit: The current state of the pull request can be tracked here: 
https://trello.com/b/91uwJaI3/autotools-milestone

### CODE REVIEW

**Comment by tomasbrod** [comment]
Woah.
I know it is not done yet. Can you explain why you had to add so many files?
Sixty six thousand lines juts to support building with autotools. This is clearly unacceptable.
Besides this is *Grid*coin, not bitcoin.

On the bright side, I like the idea with shared library.

**Review by tomasbrod — CHANGES_REQUESTED** [review]
~clean it~ dang, how do I delete this :(

**Comment by denravonska** [comment]
A lot of new files (doc, contrib etc) from Bitcoin it seems.

**Comment by gridcoin** [comment]
If this were merged in, would we still be able to do a native build on windows with the pro file, or would that break?

**Comment by denravonska** [comment]
@gridcoin We would need a new procedure but I'll try to fix that. Hopefully this removes the need for a dev vdisk.

File: src/qt/bitcoinstrings.cpp
Comment (denravonska) [inline src/qt/bitcoinstrings.cpp:5]: Do we need to regenerate this?

File: src/qt/votingdialog.cpp
Comment (denravonska) [inline src/qt/votingdialog.cpp]: Is this really the way to do it? When using qmake it automatically appends the correct include path (although with `-isystem` for some reason) so you don't have to specify the module prefix:

```
-isystem /usr/include/x86_64-linux-gnu/qt5/QtConcurrent
```

This should be the same for both Qt4 and 5 IIRC.

**Comment by sedited** [comment]
@tomasbrod I removed a bunch of the files, that were not needed and were just stuck in there, because it was easiest, when porting the build system. I started with some of the renaming, maybe a task that I can do now every evening for an hour over the next few days. I tried it with search and replace, but since bitcoin has some clearer term distinctions, I ended up with duplicate definitions. 

Currently the added files are: 

- m4 macros, Makefiles, scripts and config files for autoconf
- Edited and new source files due to creating a library and enabling support for both qt4 and qt5
- The depends system as it currently is contained in the bitcoin source code
- Hook files for gitian  
- Documentation for autotools, gitian and depends

As it currently stands, the .pro file does not work. However quite about every ide has good support for autotools and switching does not seem to be too much of a headache, or at least I had the feeling when doing it on QtCreator.

File: src/qt/bitcoin.cpp
Comment (denravonska) [inline src/qt/bitcoin.cpp]: No handling anymore? :(

Comment (sedited) [reply src/qt/bitcoin.cpp]: These were removed by @Lederstrumpf for debugging purposes. Readded them now.

File: src/qt/votingdialog.cpp
Comment (sedited) [reply src/qt/votingdialog.cpp]: The m4 for the qt checks was not properly configured. This is fixed now and the #if block is no longer required.

**Comment by denravonska** [comment]
contrib/GridcoinGalaza/GridcoinGalaza/Resources/Thumbs.db snuck back in.

**Comment by denravonska** [comment]
I get an error when configuring for a Qt which lacks Qt Charts:

```
$ ./configure --with-gui=qt5 --with-incompatible-bdb
...
checking for QT5... no
configure: error: Qt dependencies not found

$ cat config.log
...
configure:23799: checking for QT5
configure:23806: $PKG_CONFIG --exists --print-errors "$qt5_modules"
Package Qt5Charts was not found in the pkg-config search path.
Perhaps you should add the directory containing `Qt5Charts.pc'
to the PKG_CONFIG_PATH environment variable
No package 'Qt5Charts' found
configure:23809: $? = 1
configure:23823: $PKG_CONFIG --exists --print-errors "$qt5_modules"
```

**Comment by sedited** [comment]
The Qt5Charts issue is solved now, at least for Linux. I tested the build on openSUSE running Qt 5.4 and 5.9.1, and on Debian Jessie. The m4 check for the Qt Version was not properly implemented. 
Edit: Removed the Thumbs.db file again.

**Comment by ArchRoller** [comment]
It might be better to switch to CMake instead autotools.

**Comment by sedited** [comment]
The benefit of this switch to autotools is that we can use the swath of documentation, tools and support that bitcoin provides. Our source code is an old version of bitcoin to the most extent and should be updated to the current state of bitcoin step by step. I think this is a logical first step. 
As it currently stands this pull request can do what a normal cmake build could do as well. One of the tools that bitcoin uses that would be a great addition is the depends system, which uses autotools to achieve deterministic builds on windows, linux and mac. From these gitian builds can be distributed. We can get these now as well, with little additional cost. 

If you can achieve building the depends system with gitian in cmake, that would be cool. No cryptocurrency so far managed to do that and it has been tried by monero. Until then, this is the safest way to a better and tested build system.

**Review by tomasbrod — COMMENTED** [review]
This change direction is very good. But the only thing that I don't like it the shear number of added files: 131 new files.

**Comment by sedited** [comment]
I get why the number of files is troubling. A large number of them are from the m4 macros needed to check the environment and the depends system. A quick way to check their integrity would be to check if they differ against their state in the bitcoin repository, since they were largely left untouched.

**Comment by tomasbrod** [comment]
What about the images in the doc/gitian-building directory?
What is build-aux/?
And depends/?

**Comment by sedited** [comment]
`build-aux/` contains all the additional m4 files needed to check the environment. For example boost version, qt configs etc. Also, all the generated libtool files go in there. 

`depends/` is the folder containing the recipes to get and compile all the dependencies required to compile gridcoin completely from source. You can do this now for example a 64bit windows build by running `make HOST=x86_64-w64-mingw32` in the depends folder (currently this still fails because libzip does not correctly link zlib). All the dependencies are then pulled, compiled and linked in there. Configure is then pointed to the depends and built with `CONFIG_SITE=$PWD/depends/x86_64-w64-mingw32/share/config.site` . This is required for true cross platform building and gitian building. 

The images are just part of the documentation file for the gitian build process. If I do not manage to get gitian running before the pr is merge ready, I will remove these and add them back at a later point.

**Comment by sedited** [comment]
To compile the windows daemon and gui in the same fashion, I had to remove all the `WIN32 && QT_GUI` #ifdefs from the source files and put them into two distinctive sources: `neuralnet.cpp` and `grcrestarter.cpp` . Further, I created `neuralnet_stub.cpp` and `grcrestarter_stub.cpp` to contain another version of the code without the ifdefs as future points to include the new neural net / stats harvester. These new files need to be reviewed and tested.

The deamon and gui are not compiling yet on windows, but it seems close.

**Comment by sedited** [comment]
Autotools is feature complete now. I probably need some help rebasing, since the list of conflicts is quite long.

**Comment by denravonska** [comment]
Merging this to end the maintenance. Remaining issues:

- Versioning is incorrect. It says v0.14.99.0-unk.
- In OSX I get the old icon on testnet and no icon in prod. Might affect dev/ already.

**Comment by theMarix** [comment]
I know I am a little late to the party. But, there used to be a `NO_UPGRADE`  flag which enabled package maintainers to disable the upgrader as this does not serve any purpose if Gridcoin is installed from a distribution package. Sadly I can't find anything like this in `configure --help`. Is there still some way to do this or did it get lost in translation?

**Comment by sedited** [comment]
The upgrader was permanently disabled. Libzip and curl have been removed as dependencies.

**Pull request merged**
